### PR TITLE
fix(spec-decode): resync thinking state before target logits in rejec…

### LINKED ISSUE #41758: [Bug]: ngram speculative decoding changes greedy output on Qwen3-0.6B / A100
### Your current environment

<details>
<summary>Environment</summary>

```text
vLLM: 0.20.1
Docker image: vllm/vllm-openai@sha256:9eff9734a30b6713a8566217d36f8277630fd2d31cec7f0a0292835901a23aa4
Model: Qwen/Qwen3-0.6B
Model revision: c1899de289a04d12100db370d81485cdf75e47ca
GPU: NVIDIA A100-SXM4-40GB
Driver: 580.105.08
CUDA: 13.0

Attached bundle includes:
- environment/nvidia_smi.txt
- environment/docker_image_inspect.json
- model_snapshot.json
- server logs
```

</details>

### 🐛 Describe the bug

I found a deterministic output difference between a no-speculative vLLM server and the same image/config with ngram speculative decoding enabled.

Summary:

- vLLM: `0.20.1`
- Docker image: `vllm/vllm-openai@sha256:9eff9734a30b6713a8566217d36f8277630fd2d31cec7f0a0292835901a23aa4`
- Model: `Qwen/Qwen3-0.6B`
- Model revision: `c1899de289a04d12100db370d81485cdf75e47ca`
- GPU: `NVIDIA A100-SXM4-40GB`
- Driver: `580.105.08`
- CUDA: `13.0`
- Params: `temperature=0`, `top_p=1`, `seed=42`, `max_tokens=32`
- Prompt: `"serving" * 48`
- Prompt SHA-256: `f11888cdb5376f51784f3b231beaa61a69faac2d068ee29dcb7f002f79a0a5fb`
- Baseline: same vLLM server without speculative decoding
- Feature server: `--speculative-config '{"method":"ngram","num_speculative_tokens":4,"prompt_lookup_min":2,"prompt_lookup_max":5}'`

Expected:

With greedy decoding and the same model, prompt, seed, and sampling parameters, ngram speculative decoding should return the same final generated text as the no-speculative baseline.

Observed:

```text
baseline:
'\nOkay, let\'s see. The user provided a string of "servingservingservingservingservingservingservingservingservings'

ngram speculative:
'\nOkay, let\'s see. The user has provided a string of "servingservingservingservingservingservingservingservingserv'
```

Confirmation:

- repeated `serving` prompt: `20/20` baseline-vs-ngram comparisons mismatched
- code-like control prompt with the same servers/params: `20/20` comparisons matched

I found a few related speculative/ngram issues such as #16883 and #11123, but this repro is a fixed-seed greedy baseline-vs-ngram output mismatch on vLLM 0.20.1 with a single prompt.

Attached bundle includes:

- `README.md`
- `repro.py`
- `docker-compose.yml`
- `trace.json`
- `terminal_demo.txt`
- request/response JSONL for the failing prompt
- server logs
- Docker image inspect and `nvidia-smi` snapshot
- model revision and config/tokenizer hashes

I did not dig into root cause here; the bundle is just meant to make the behavior easy to replay.

[vllm_spec_ngram_qwen3_0_6b_a100.zip](https://github.com/user-attachments/files/27415060/vllm_spec_ngram_qwen3_0_6b_a100.zip)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by thomascolden585-svg**
## Related

- Closes / fixes: https://github.com/vllm-project/vllm/issues/41758 (greedy output differed between baseline and `method: ngram` speculative decoding under the same seed and sampling params; repro included Qwen3-0.6B and repeated-token prompts).

## Purpose

1. **Spec decode + reasoning / thinking budget:** In `RejectionSampler.forward`, the bonus-token path runs `Sampler` first, which calls `ThinkingBudgetStateHolder.update_state` and `apply_to_logits(..., predict_bonus_token=True)`. That mutates forcing state (`force_index`, `bonus_token_forced`). Target draft logits were then processed with `apply_to_logits(..., predict_bonus_token=False)` **without** a matching `update_state`, so stale forcing could change masked logits and **greedy argmax** vs plain (non-spec) decode.

2. **`min_tokens` + spec decode:** `MinTokensLogitsProcessor.apply_with_spec_decode` masked EOS on one **extra** draft row (off-by-one vs “output length after this row is still `< min_tokens`”).

3. **Batch alignment:** `RejectionSampler._combine_outputs_with_spec_tokens` skipped requests with empty `spec_token_ids`, which could misalign rows when mixing requests in one batch.

## Changes (summary)

- `vllm/v1/sample/rejection_sampler.py`: Call `holder.update_state(..., repeat_indices)` before target `apply_to_logits`; fix `_combine_outputs_with_spec_tokens` for empty spec lists.
- `vllm/v1/sample/logits_processor/builtin.py`: Fix `n_mask` using `min_tok - current_len - 1`.

## Duplicate work check (per AGENTS.md)

Before opening the PR, confirm no open PR already fixes the same issue:

```bash
gh issue view 41758 --repo vllm-project/vllm --comments
gh pr list --repo vllm-project/vllm --state open --search "41758 in:body"
gh pr list --repo vllm-project/vllm --state open --search "ngram speculative greedy"
```

If another PR is in flight, coordinate there instead of duplicating.

## Test plan

```bash
# Lint (after uv venv + pre-commit install)
pre-commit run --files vllm/v1/sample/rejection_sampler.py vllm/v1/sample/logits_processor/builtin.py

# Targeted tests
.venv/bin/python -m pytest tests/v1/spec_decode/test_ngram.py -v
.venv/bin/python -m pytest tests/v1/logits_processors/test_correctness.py -v
```

Optional (if reasoning + OpenAI path is available):

```bash
.venv/bin/python -m pytest tests/entrypoints/openai/chat_completion/test_thinking_token_budget.py -v
```

## Test result

_Paste output here after running the above (pass/fail, environment: GPU/CPU, vLLM commit)._

## AI assistance

This change was developed with AI assistance. A human submitter must review every line, run the tests above, and defend the change in review.

## Checklist (PR template)

- [x] Purpose tied to issue #41758 and root cause described.
- [x] Test plan with concrete commands.
- [ ] Test results pasted (runner fills in).
- [ ] Duplicate PR search done (runner fills in).
- [ ] No unrelated drive-by refactors.

### CODE REVIEW

**Comment by rishitdholakia13** [comment]
> Spec decode + reasoning / thinking budget: In RejectionSampler.forward, the bonus-token path runs Sampler first, which calls ThinkingBudgetStateHolder.update_state and apply_to_logits(..., predict_bonus_token=True). That mutates forcing state (force_index, bonus_token_forced). Target draft logits were then processed with apply_to_logits(..., predict_bonus_token=False) without a matching update_state, so stale forcing could change masked logits and greedy argmax vs plain (non-spec) decode.

Hi, based on the issue reported [here](https://github.com/vllm-project/vllm/issues/41758)
, it appears the engine is not currently applying the thinking budget. Since no thinking budget is assigned, that branch of the code is effectively skipped.

We also evaluated the change you’re suggesting earlier. With thinking budget combined with speculative decoding, updating the state twice results in incorrect thinking budget calculations at that step. In particular, the state update during bonus token sampling already accounts for both the target logits and draft tokens, ensuring the target’s logits processor does not operate on stale state. Given this, the additional changes introduced in this PR can likely be removed safely.

Could you share a concrete example  where the current behavior leads to the model overbudgeting ?
The test_thinking_token_budget, contains non-Sd as well as SD tests, and they do pass with the right budgets. If we are to use the update_state again for the target logits, you would notice that the test breaks.
